common/djangoapps/auth_exchange/views.py

```diff
@@ -44,12 +62,99 @@ def post(self, request, _backend):  # pylint: disable=arguments-differ
         scope = form.cleaned_data["scope"]
         client = form.cleaned_data["client"]
 
+        return self.exchange_access_token(request, user, scope, client)
+
+    def exchange_access_token(self, request, user, scope, client):
+        """
+        Exchange third party credentials for an edx access token, and return a
+        serialized access token response.
+        """
         if constants.SINGLE_ACCESS_TOKEN:
-            edx_access_token = self.get_access_token(request, user, scope, client)
+            edx_access_token = self.get_access_token(request, user, scope, client)  # pylint: disable=no-member
         else:
             edx_access_token = self.create_access_token(request, user, scope, client)
+        return self.access_token_response(edx_access_token)  # pylint: disable=no-member
+
+
+class DOPAccessTokenExchangeView(AccessTokenExchangeBase, DOPAccessTokenView):
+    """
+    View for token exchange from 3rd party OAuth access token to 1st party
+    OAuth access token.  Uses django-oauth2-provider (DOP) to manage access
+    tokens.
+    """
+
+    oauth2_adapter = adapters.DOPAdapter()
+
+
+class DOTAccessTokenExchangeView(AccessTokenExchangeBase, DOTAccessTokenView):
+    """
+    View for token exchange from 3rd party OAuth access token to 1st party
+    OAuth access token.  Uses django-oauth-toolkit (DOT) to manage access
+    tokens.
+    """
+
+    oauth2_adapter = adapters.DOTAdapter()
+
+    def get(self, request, _backend):
+        return Response(status=400, data={
+            'error': 'invalid_request',
+            'error_description': 'Only POST requests allowed.',
+        })
+
+    def get_access_token(self, request, user, scope, client):
+        """
+        Return an access token, using an existing token, if available.
+        """
+        try:
+            return dot_models.AccessToken.objects.get(
+                user=user,
+                application=client,
+                scope=scope,
+                expires__gt=datetime.utcnow(),
+            )
+        except dot_models.AccessToken.DoesNotExist:
+            return self.create_access_token(request, user, scope, client)
+
+    def create_access_token(self, request, user, scope, client):
+        """
+        Create and return a new access token.
+        """
+        _days = 24 * 60 * 60
+        token_generator = BearerToken(
+            expires_in=settings.OAUTH_EXPIRE_PUBLIC_CLIENT_DAYS * _days,
+            request_validator=OAuth2Validator(),
```

Comment:
This is fine for now.  Just wondering, does DOT expect the validator to be configurable?  So the caller can specify/override which one to use?